packages/contracts-sdk/src/implementations/farcaster/FarcasterKeyGatewayContract.ts

```diff
@@ -0,0 +1,248 @@
+  ): ResultAsync<
+    SignedAddSignature,
+    FarcasterKeyGatewayContractError | BlockchainCommonErrors
+  > {
+    return this.nonces(ownerAddress).andThen((latestNonce) => {
+      const addKeyMessage = {
+        owner: ownerAddress,
+        keyType: 1,
+        key: keyToAdd,
+        metadataType: 1,
+        metadata: encodedMetadata,
+        nonce: latestNonce,
+        deadline: deadline,
+      };
+
+      const signer = this.providerOrSigner as ethers.Signer;
+
+      return ResultAsync.fromPromise(
+        signer.signTypedData(
+          KEY_GATEWAY_EIP_712_TYPES.domain,
+          this.removeReadonlyFromReadonlyTypes(KEY_GATEWAY_EIP_712_TYPES.types),
+          addKeyMessage,
+        ) as Promise<SignedAddSignature>,
+        (e) => {
+          return e as FarcasterKeyGatewayContractError;
+        },
+      );
+    });
+  }
+
+  public getSignedKeyRequestSignatureAndEncodedMetadata(
+    ownerFid: FarcasterId,
+    keyToAdd: EVMAccountAddress, // key to be tied to account
+    deadline: UnixTimestamp,
+  ): ResultAsync<SignedKeyRequest, FarcasterKeyGatewayContractError> {
+    return this.getSignedKeyRequestSignature(
+      ownerFid,
+      keyToAdd,
+      deadline,
+    ).andThen((metadataSignature) => {
+      const metadataStruct = {
+        requestFid: ownerFid,
+        requestSigner: keyToAdd,
+        signature: metadataSignature,
+        deadline: deadline,
+      };
+
+      const metadataStructType = ParamType.from({
+        components: [
+          {
+            name: "requestFid",
+            type: "uint256",
+          },
+          {
+            name: "requestSigner",
+            type: "address",
+          },
+          {
+            name: "signature",
+            type: "bytes",
+          },
+          {
+            name: "deadline",
+            type: "uint256",
+          },
+        ],
+        name: "SignedKeyRequestMetadata",
+        type: "tuple",
+      });
+
+      const encodedMetadataStruct = ethers.AbiCoder.defaultAbiCoder().encode(
+        [metadataStructType],
+        [metadataStruct],
+      );
+
+      return okAsync(
+        new SignedKeyRequest(
+          SignedKeyRequestSignature(metadataSignature),
+          EncodedSignedKeyRequestMetadata(encodedMetadataStruct),
+        ),
+      );
+    });
+  }
+
+  // Returns SignedKeyRequestSignature
+  public getSignedKeyRequestSignature(
+    ownerFid: FarcasterId,
+    keyToAdd: EVMAccountAddress, // key to be tied to account
+    deadline: UnixTimestamp,
+  ): ResultAsync<SignedKeyRequestSignature, FarcasterKeyGatewayContractError> {
+    //    https://docs.farcaster.xyz/reference/contracts/reference/signed-key-request-validator#signedkeyrequestmetadata-struct
+    //    Needs to be a EIP712 signature
+    const metadataMessage = {
+      requestFid: ownerFid,
+      key: keyToAdd,
+      deadline,
+    };
+
+    // Use the wallet of the sdk caller as the signer
+    const signer = this.providerOrSigner as ethers.Signer;
```

Comment:
Don't we have a method in the base contract that will assure this is a signer and error if not?